File: packages/sdk-ts/examples/webmcp.ts
import { localBrowser, Stagehand } from "../src/index.js";

const webMCPTestSite = "https://browserbase.github.io/stagehand-eval-sites/sites/webmcp-test/";

const browser = await localBrowser.launch({ headless: false });
const stagehand = await Stagehand.create({ browser });

try {
  const [page] = await browser.context.pages();
  // Subscribe before navigation: hooks report future changes, not existing tools.
  const added = await page.onToolsAdded((tools) => {
    for (const tool of tools) console.log("Tool added:", tool.name, tool.frameId);
  });
  let markRemovalReceived!: () => void;
  const removalReceived = new Promise<void>((resolve) => {
    markRemovalReceived = resolve;
  });
  const removed = await page.onToolsRemoved((tools) => {
    for (const tool of tools) console.log("Tool removed:", tool.name, tool.frameId);
    if (tools.length > 0) markRemovalReceived();
  });
  await page.goto(webMCPTestSite);

  const tools = await page.tools({ timeout: 5_000 });
  const calculateSum = tools.find((tool) => tool.name === "calculateSum");
  if (!calculateSum) {
    throw new Error("calculateSum was not registered by the page");
  }

  const invocation = await calculateSum.invoke({
    input: { a: 19, b: 23 },
  });
  const result = await invocation.result();

  console.log(result);

  // Leaving the document removes its registered tools.
  await page.goto("about:blank");
  let timer: ReturnType<typeof setTimeout> | undefined;
  try {
    await Promise.race([
      removalReceived,
      new Promise<never>((_, reject) => {
        timer = setTimeout(() => reject(new Error("Timed out waiting for tool removal")), 5_000);
      }),
    ]);
  } finally {
    clearTimeout(timer);
  }
  await added.unsubscribe();
  await removed.unsubscribe();
} finally {
  try {
    await stagehand.close();
  } finally {
    await browser.close();
  }
}

File: packages/sdk-python/examples/webmcp.py
import asyncio

from stagehand import Stagehand, WebMCPTool, WebMCPToolIdentity, local_browser

WEBMCP_TEST_SITE = "https://browserbase.github.io/stagehand-eval-sites/sites/webmcp-test/"


def tools_added(tools: list[WebMCPTool]) -> None:
    for tool in tools:
        print("Tool added:", tool.name, tool.frame_id)


async def main() -> None:
    removal_received = asyncio.Event()

    def tools_removed(tools: list[WebMCPToolIdentity]) -> None:
        for tool in tools:
            print("Tool removed:", tool.name, tool.frame_id)
        if tools:
            removal_received.set()

    browser = await local_browser.launch(headless=False)
    try:
        stagehand = await Stagehand.create(browser=browser)
        try:
            page = (await browser.context.pages())[0]
            if page is None:
                raise RuntimeError("Stagehand initialized without an active page")
            # Subscribe before navigation: hooks report future changes, not existing tools.
            added = await page.on_tools_added(tools_added)
            removed = await page.on_tools_removed(tools_removed)
            await page.goto(WEBMCP_TEST_SITE)

            tools = await page.tools(timeout=5_000)
            calculate_sum = next(
                (tool for tool in tools if tool.name == "calculateSum"),
                None,
            )
            if calculate_sum is None:
                raise RuntimeError("calculateSum was not registered by the page")

            invocation = await calculate_sum.invoke(input={"a": 19, "b": 23})
            result = await invocation.result()

            print(result.model_dump_json(indent=2))

            # Leaving the document removes its registered tools.
            await page.goto("about:blank")
            await asyncio.wait_for(removal_received.wait(), timeout=5)
            await added.unsubscribe()
            await removed.unsubscribe()
        finally:
            await stagehand.close()
    finally:
        await browser.close()


if __name__ == "__main__":
    asyncio.run(main())
